usr/lib/common/pkcs_utils.c

```diff
@@ -0,0 +1,448 @@
+#include "host_defs.h"
+
+#define OCK_TOOL
+#include "pkcs_utils.h"
+
+extern pkcs_trace_level_t trace_level;
+
+void pkcs_trace(pkcs_trace_level_t level, const char *file, int line,
+                const char *fmt, ...)
+{
+    va_list ap;
+    const char *fmt_pre;
+    char buf[1024];
+    char *pbuf;
+    int buflen, len;
+
+    if (level > trace_level)
+        return;
+
+    pbuf = buf;
+    buflen = sizeof(buf);
+
+    /* add file line */
+    switch (level) {
+    case TRACE_LEVEL_NONE:
+        fmt_pre = "";
+        break;
+    case TRACE_LEVEL_ERROR:
+        fmt_pre = "[%s:%d] ERROR: ";
+        break;
+    case TRACE_LEVEL_WARNING:
+        fmt_pre = "[%s:%d] WARN: ";
+        break;
+    case TRACE_LEVEL_INFO:
+        fmt_pre = "[%s:%d] INFO: ";
+        break;
+    case TRACE_LEVEL_DEVEL:
+        fmt_pre = "[%s:%d] DEVEL: ";
+        break;
+    case TRACE_LEVEL_DEBUG:
+        fmt_pre = "[%s:%d] DEBUG: ";
+        break;
+    default:
+        return;
+    }
+    snprintf(pbuf, buflen, fmt_pre, file, line);
+
+    len = strlen(buf);
+    pbuf = buf + len;
+    buflen = sizeof(buf) - len;
+
+    va_start(ap, fmt);
+    vsnprintf(pbuf, buflen, fmt, ap);
+    va_end(ap);
+
+    printf("%s", buf);
+}
+
+int compute_hash(int hash_type, int buf_size, char *buf, char *digest)
+{
+    EVP_MD_CTX *md_ctx = NULL;
+    unsigned int result_size;
+    int rc;
+
+    md_ctx = EVP_MD_CTX_create();
+
+    switch (hash_type) {
+    case HASH_SHA1:
+        rc = EVP_DigestInit(md_ctx, EVP_sha1());
+        break;
+    case HASH_MD5:
+        rc = EVP_DigestInit(md_ctx, EVP_md5());
+        break;
+    default:
+        EVP_MD_CTX_destroy(md_ctx);
+        return -1;
+        break;
+    }
+
+    if (rc != 1) {
+        TRACE_ERROR("EVP_DigestInit() failed: rc = %d\n", rc);
+        return -1;
+    }
+
+    rc = EVP_DigestUpdate(md_ctx, buf, buf_size);
+    if (rc != 1) {
+    	TRACE_ERROR("EVP_DigestUpdate() failed: rc = %d\n", rc);
+        return -1;
+    }
+
+    result_size = EVP_MD_CTX_size(md_ctx);
+    rc = EVP_DigestFinal(md_ctx, (unsigned char *) digest, &result_size);
+    if (rc != 1) {
+    	TRACE_ERROR("EVP_DigestFinal() failed: rc = %d\n", rc);
+        return -1;
+    }
+
+    EVP_MD_CTX_destroy(md_ctx);
+
+    return 0;
+}
+
+CK_RV local_rng(CK_BYTE *output, CK_ULONG bytes)
+{
+    int ranfd;
+    int rlen;
+    unsigned int totallen = 0;
+
+    ranfd = open("/dev/prandom", 0);
+    if (ranfd < 0)
+        ranfd = open("/dev/urandom", 0);
+    if (ranfd >= 0) {
+        do {
+            rlen = read(ranfd, output + totallen, bytes - totallen);
+            totallen += rlen;
+        } while (totallen < bytes);
+        close(ranfd);
+        return CKR_OK;
+    }
+
+    return CKR_FUNCTION_FAILED;
+}
+
+CK_RV aes_256_wrap(unsigned char out[40], const unsigned char in[32],
+                   const unsigned char kek[32])
+{
+    CK_RV rc;
+    int outlen;
+    unsigned int buffer[40 + 16]; // add 1 block for openssl internal usage
```

Comment:
same comments here as in your other commit

Reply:
see above.